File: streams_and_processes/3_filter_programs/Double.java
/*

*/

import java.util.Scanner;

/**
   This program reads one line at a time from standard input,
   and then writes each character twice to standard output.
<p>
   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).

   @see DoubleN
*/
public class Double
{
   public static void main(String[] args)
   {
      // Create a Scanner object to make it easier to use System.in
      final Scanner scanner = new Scanner( System.in );

      while ( scanner.hasNextLine() )
      {
         final String oneLine = scanner.nextLine();
         for (int i = 0; i < oneLine.length(); i++)
         {
            System.out.print( oneLine.charAt(i) );
            System.out.print( oneLine.charAt(i) );
         }
         System.out.println();
         if ( System.out.checkError() )
            throw new RuntimeException("System.out has encountered an IOException");
      }
   }


   // Private default constructor to enforce noninstantiable class.
   // See Item 4 in "Effective Java", 3rd Ed, Joshua Bloch.
   private Double() {
      throw new AssertionError();
   }
}

File: streams_and_processes/5_creating_a_pipe/Windows/double.c
/*
   This program reads one character at a time from standard input,
   and then writes each character twice to standard output.

   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).
*/
#include <stdio.h>

int main()
{
   int c;
   while ( (c = getchar()) != EOF )
   {
      if ( c != 10 && c != 13 )
      {
         printf("%c%c", c, c);
      }
      else
      {
         printf("%c", c);
      }
   }
   return 0;
}
